### PR TITLE
skip already-flushed events when replaying after stream recreation

### PR DESCRIPTION
**Pull request by morgo**
## Problem

After 5 consecutive stream errors, `recreateStreamer` restarts the binlog dump at position 4 of the current file, and `readStream` re-delivered every replayed RowsEvent to subscriptions. A replayed event at or below `flushedPos` can regress a key in the buffered map to a stale row image. Since `setBufferedPos` is monotonic (#847), the periodic flush is free to run mid-replay and write that stale image to the target — below the already-persisted checkpoint:

1. Events E1 (key K, ends pos 100) and E2 (key K, ends pos 200) are read; map holds K@200.
2. Flush applies K@200; `flushedPos`=200; checkpoint persisted.
3. Stream errors → replay from position 4 re-reads E1: map now holds the stale K@100.
4. Periodic flush applies K@100 to the target.
5. Crash before the replay re-reads E2. Resume streams only events after pos 200 — E2 is never re-sent, and the row keeps the older image permanently. With `--checksum=false` there is no backstop.

Cutover via `FlushUnderTableLock` is protected (BlockWait forces the replay to finish); only the crash/resume path is exposed. The GTID client does not share this bug (it resumes from fully-buffered committed transactions).

## Fix

`recreateStreamer` captures the current `flushedPos` on every recreation, and `readStream` skips delivering replayed RowsEvents whose end position is at or below it — those events are durably applied per the `Position()` contract. Events above it are re-delivered as before (idempotent re-buffering). Comparison uses full `(file, pos)` ordering, so rotation is handled. `Start` resets the filter.

## Testing

- `TestShouldSkipReplayedEvent`: table-driven unit tests for the skip decision (zero filter, below/equal/above, earlier/later file).
- `TestRecreateStreamerSkipsFlushedReplay`: integration test that kills the syncer so the genuine consecutive-error path drives `recreateStreamer`, then asserts flushed events are NOT re-buffered while unflushed ones are, and that a second recreation re-captures the advanced filter. Fails with delta 2 vs expected 1 when the skip is disabled.
- Full `pkg/change` suite passes with `-race`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pkg/change/binlog.go [hunk 1]
```diff
@@ -179,6 +198,39 @@ func (c *binlogClient) getBufferedPos() mysql.Position {
 	return c.bufferedPos
 }
 
+// getReplayFilterPos returns the replay filter position under a mutex.
+// See the replayFilterPos field for semantics.
+func (c *binlogClient) getReplayFilterPos() mysql.Position {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.replayFilterPos
+}
```
[the remark this answers is not in the document]
Comment (morgo) [reply pkg/change/binlog.go]: Good catch — this is a real gap, and the fix went further than the literal suggestion.

You are right that a snapshot of `flushedPos` is insufficient: the periodic flush (every 30s) runs concurrently with the replay and advances `flushedPos` all the way to the frozen `bufferedPos` (`setBufferedPos` is monotonic, so replayed events do not move it). A snapshot taken at recreate time leaves the `(snapshot, bufferedPos]` window unfiltered, a replayed event there overwrites a key in the buffered map with a stale image, and the next flush persists it below the now-advanced checkpoint — permanent after a crash/resume.

However, widening to "any later flushedPos once replay filtering is active" (a live `flushedPos` read) does **not** close the window either. The corruption happens at the moment a replayed event in `(flushedPos, bufferedPos]` is delivered — `flushedPos` has not advanced yet at that instant, so a live-`flushedPos` comparison still lets it through and corrupts the map; the subsequent flush then persists the stale image. Trace: map={K:v3}, flushedPos=100, bufferedPos=300; replay delivers E2@200 (200>100, map→{K:v2}), flush sets flushedPos=300 persisting v2, E3@300 then skipped — K stuck at the stale v2.

The correct boundary is the **live `bufferedPos`**, not `flushedPos`. Every event at or below `bufferedPos` is already faithfully represented — durably applied (<= flushedPos) **or** held in the map with its latest deduped image (in `(flushedPos, bufferedPos]`) — so skipping its re-delivery can never corrupt the map. `bufferedPos` is frozen during replay and only advances when a genuinely-new event past the high-water mark is delivered, so the boundary widens with the stream and the filter is naturally inert for live traffic.

Changes (commit 8f6b897f):
- `recreateStreamer` now sets a `replayActive` flag (arms the filter) instead of snapshotting `flushedPos`; `replayFilterFloor` keeps the recreate-time `bufferedPos` for diagnostics only.
- `shouldSkipReplayedEvent(replayActive, eventPos, bufferedPos)` compares each replayed event against the live `bufferedPos` read under `mu`.
- Unit tests extended for the advanced-boundary cases the snapshot missed and the boundary-widening property; the integration test now buffers the unflushed event before recreation so it exercises the `(flushedPos, bufferedPos]` window.

`go test ./pkg/change/... -count=1 -race` green; `go build ./...`, `go vet`, and `golangci-lint` clean.

**Review by aparajon — APPROVED** [review]
